drivers/rtc/rtc_counter.c

```diff
@@ -0,0 +1,464 @@
+		memset(timeptr, 0, sizeof(struct rtc_time));
+		return;
+	}
+
+	timeptr->tm_sec = tm_val.tm_sec;
+	timeptr->tm_min = tm_val.tm_min;
+	timeptr->tm_hour = tm_val.tm_hour;
+	timeptr->tm_mday = tm_val.tm_mday;
+	timeptr->tm_mon = tm_val.tm_mon;
+	timeptr->tm_year = tm_val.tm_year;
+	timeptr->tm_wday = tm_val.tm_wday;
+	timeptr->tm_yday = tm_val.tm_yday;
+	timeptr->tm_isdst = -1;
+	timeptr->tm_nsec = 0;
+}
+
+static int counter_rtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct counter_rtc_config *config = dev->config;
+	struct counter_rtc_data *data = dev->data;
+	uint32_t desired_ticks = 0;
+	uint32_t now_ticks = 0;
+	int ret;
+
+	if (timeptr == NULL) {
+		return -EINVAL;
+	}
+
+	ret = counter_rtc_time_to_ticks(timeptr, &desired_ticks);
+
+	/* -EINVAL on overflow/invalid time */
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Stop counter */
+	ret = counter_stop(config->counter_dev);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = counter_get_value(config->counter_dev, &now_ticks);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Update the software offset: offset = desired_time - current_raw_ticks */
+	K_SPINLOCK(&data->lock) {
+		data->epoch_offset = (int64_t)desired_ticks - (int64_t)now_ticks;
+	}
+
+	/* Restart counter */
+	return counter_start(config->counter_dev);
+}
+
+static int counter_rtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	const struct counter_rtc_config *config = dev->config;
+	struct counter_rtc_data *data = dev->data;
+	uint32_t now_ticks;
+	int ret;
+	int64_t epoch;
+	int64_t current_seconds;
+
+	if (timeptr == NULL) {
+		return -EINVAL;
+	}
+
+	ret = counter_get_value(config->counter_dev, &now_ticks);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		epoch = data->epoch_offset;
+	}
+
+	current_seconds = (int64_t)now_ticks + epoch;
+
+	if (current_seconds < 0 || current_seconds > UINT32_MAX) {
+		return -ERANGE;
+	}
+
+	memset(timeptr, 0, sizeof(struct rtc_time));
+	counter_rtc_ticks_to_time((uint32_t)current_seconds, timeptr);
+
+	return 0;
+}
+
+#if defined(CONFIG_RTC_ALARM)
+
+static int counter_rtc_alarm_get_supported_fields(const struct device *dev, uint16_t id,
+						  uint16_t *mask)
+{
+	struct counter_rtc_data *data = dev->data;
+
+	if (!data->alarm_capable) {
+		return -ENOTSUP;
+	}
+
+	if (id >= data->num_alarm_chans) {
+		return -EINVAL;
+	}
+
+	*mask = (RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE |
+		 RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MONTHDAY |
+		 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_YEAR |
+		 RTC_ALARM_TIME_MASK_WEEKDAY | RTC_ALARM_TIME_MASK_YEARDAY);
+
+	return 0;
+}
+
+static void counter_rtc_alarm_callback(const struct device *counter_dev, uint8_t chan_id,
+				       uint32_t ticks, void *user_data)
+{
+	struct counter_rtc_data *data = (struct counter_rtc_data *)user_data;
+	const struct device *rtc_dev = data->rtc_dev;
+
+	if (chan_id < data->num_alarm_chans && data->alarm_callback[chan_id]) {
```

Comment:
Fix implicit boolean conversions

```suggestion
	if (chan_id < data->num_alarm_chans && data->alarm_callback[chan_id] != NULL) {
```

Reply:
Updated.